src/coreclr/jit/morph.cpp

```diff
@@ -14393,6 +14393,14 @@ void Compiler::fgMorphTreeDone(GenTree* tree,
     /* If this tree makes a new assertion - make it available */
     optAssertionGen(tree);
 
+    // For struct copies & inits, also use the original tree
+    // to generate assertions.
+    //
+    if ((oldTree != nullptr) && oldTree->OperIsBlkOp())
+    {
+        optAssertionGen(oldTree);
+    }
```

Comment:
I don't think we can really use `oldTree` here... It should really be a `DEBUGARG`, the contract is that it can be reshapen/destroyed/reused in arbitrary ways.

Reply:
All we really need to know is the identity of the local(s) involved, so I suppose we can find some other way to convey this information.

Reply:
Would calling this somewhere from somewhere inside `fgMorphCopy/InitBlock` not work? We are already doing this "manual" assertion generation for field assigns there.

Reply:
It needs to survive `fgKillDependentAssertions` somehow.

Reply:
Do it after the killing?

Reply:
Well sure, but that happens in `fgMorphTreeDone`, and by then we've possibly lost the orginal trees, as you point out. So, we need to capture the info early but not act on it until later, and morph isn't well set up for this sort of thing.

Reply:
That might work. It scatters the gen/kill logic around a bit more, but I suppose we've already crossed that bridge.

I certainly want to avoid having to inspect the expanded tree to figure out what to kill/preserve/gen.

Reply:
Strange that we kill assertions for init block but not for copy block.

Reply:
Ah, I guess we inherit that.

Reply:
Tried a variant of what's discussed above: putting the assertion gen over in `MorphInitBlockHelper::Morph` after the dest is prepared but before the copy is morphed.

This misses some cases where we turn the copy into a single assignment to its field.
```
;;; current PR ("hacky revival of old tree")

Morphing BB01 of 'SocketConnection:DisposeAsync():ValueTask:this'

fgMorphTree BB01, STMT00008 (before)
               [000037] IA---------                         *  ASG       struct (init)
               [000034] D------N---                         +--*  LCL_VAR   struct<System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder, 8>(P) V04 tmp1         
                                                            +--*    ref    V04.m_task (offs=0x00) -> V09 tmp6         
               [000036] -----------                         \--*  CNS_INT   int    0
Notify VM instruction set (SSE2) must be supported.
MorphInitBlock:
MorphBlock for dst tree, before:
               [000034] D----+-N---                         *  LCL_VAR   struct<System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder, 8>(P) V04 tmp1         
                                                            *    ref    V04.m_task (offs=0x00) -> V09 tmp6         
MorphBlock after:
               [000034] D----+-N---                         *  LCL_VAR   struct<System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder, 8>(P) V04 tmp1         
                                                            *    ref    V04.m_task (offs=0x00) -> V09 tmp6         
PrepareDst for [000034] have found a local var V04.
 using field by field initialization.
GenTreeNode creates assertion:
               [000047] -A---------                         *  ASG       ref   
In BB01 New Local Constant Assertion: V09 == null, index = #01
MorphInitBlock (after):
               [000047] -A---+-----                         *  ASG       ref   
               [000045] D------N---                         +--*  LCL_VAR   ref    V09 tmp6         
               [000046] -----------                         \--*  CNS_INT   ref    null

The assignment [000047] using V09 removes: Constant Assertion: V09 == null
GenTreeNode creates assertion:
               [000047] -A---+-----                         *  ASG       ref   
In BB01 New Local Constant Assertion: V09 == null, index = #01
GenTreeNode creates assertion:
               [000037] IA---------                         *  ASG       struct (init)
In BB01 New Local Constant Assertion: V04 == ZeroObj, index = #02
```
compared to
```
;;; alternate version (revive during morph block init)

Morphing BB01 of 'SocketConnection:DisposeAsync():ValueTask:this'

fgMorphTree BB01, STMT00008 (before)
               [000037] IA---------                         *  ASG       struct (init)
               [000034] D------N---                         +--*  LCL_VAR   struct<System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder, 8>(P) V04 tmp1         
                                                            +--*    ref    V04.m_task (offs=0x00) -> V09 tmp6         
               [000036] -----------                         \--*  CNS_INT   int    0
Notify VM instruction set (SSE2) must be supported.
MorphInitBlock:
MorphBlock for dst tree, before:
               [000034] D----+-N---                         *  LCL_VAR   struct<System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder, 8>(P) V04 tmp1         
                                                            *    ref    V04.m_task (offs=0x00) -> V09 tmp6         
MorphBlock after:
               [000034] D----+-N---                         *  LCL_VAR   struct<System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder, 8>(P) V04 tmp1         
                                                            *    ref    V04.m_task (offs=0x00) -> V09 tmp6         
PrepareDst for [000034] have found a local var V04.
GenTreeNode creates assertion:
               [000037] IA---------                         *  ASG       struct (init)
In BB01 New Local Constant Assertion: V04 == ZeroObj, index = #01
 using field by field initialization.
GenTreeNode creates assertion:
               [000047] -A---------                         *  ASG       ref   
In BB01 New Local Constant Assertion: V09 == null, index = #02
MorphInitBlock (after):
               [000047] -A---+-----                         *  ASG       ref   
               [000045] D------N---                         +--*  LCL_VAR   ref    V09 tmp6         
               [000046] -----------                         \--*  CNS_INT   ref    null

The assignment [000047] using V09 removes: Constant Assertion: V09 == null

The assignment [000047] using V04 removes: Constant Assertion: V04 == ZeroObj
GenTreeNode creates assertion:
               [000047] -A---+-----                         *  ASG       ref   
In BB01 New Local Constant Assertion: V09 == null, index = #01
```
I could either try and recognize those cases in `fgMorphTreeDone` and avoid the kill, or update `optAssertionGen` to understand that assertions about an entire field of a struct also apply to the struct, so that the assertion gets killed but then gets revived again.